### ISSUE TITLE
[Feature Request] Improve ease of setting dictionary priority

### ISSUE THREAD

**Issue by Thermospore** [FooSoft/yomichan#1539]
The current system of typing in a priority number works great for smaller numbers of dictionaries, but once you have more than a handful it becomes difficult to manage. For example if you have 20 dictionaries/freq lists and want to add one to the middle of the stack, you have to scroll around and bump up each of the top 10 one by one

I recall the titles used to sort themselves by priority on the settings page, which made it a bit easier. It was still a hassle to manually bump around each priority number though

Anyone have ideas of how to improve it? Drag to sort, maybe?

**Comment by toasted-nutbread** [comment]
Having the order of dictionaries be more configurable would probably fix the issue (either drag-and-drop or an option to move up/down), but there are currently some implementation issues that make this difficult without some prior rework.

> I recall the titles used to sort themselves by priority on the settings page, which made it a bit easier.

Yes, but IMO this isn't a great UX, as it's jarring if a UI element jumps around on the page after editing a text input field.

**Comment by Thermospore** [comment]
> Yes, but IMO this isn't a great UX, as it's jarring if a UI element jumps around on the page after editing a text input field.

Hmm yeah I recall that too. What if the position didn't update until the next time the menu was opened?

**Comment by toasted-nutbread** [comment]
I've made some changes to how dictionary options are presented in #1640, and added the ability to rearrange the options in #1641. Once this is in a release, let me know if that's helpful for this request.

**Comment by Thermospore** [comment]
Sure thing, thanks!!

**Comment by Thermospore** [comment]
New compact style looks nice 🙂 

possible bug:
I had prefix wildcards enabled for my dicts but the UI shows them disabled now
![image](https://user-images.githubusercontent.com/35350819/116803736-c5942900-aac6-11eb-8ea8-c21020ed1577.png)

seems they are still enabled internally
![image](https://user-images.githubusercontent.com/35350819/116803741-d04ebe00-aac6-11eb-95c2-d9ff51407946.png)

I think there are currently two concerns:
1. you can't rearrange a dictionary to it's final location in the list with a single action; if you want to move it `n` spots over you have to press the move button `n` times
2. you still need to perform some manual numbering gymnastics to reorder the priorities

perhaps something like how you rearrange the fields of a card type in Anki would be an easy solution?
![image](https://user-images.githubusercontent.com/35350819/116804211-8adfc000-aac9-11eb-9378-aba4da175428.png)
(ie you just type in the desired position and the list is automatically rearranged and renumbered)

**Comment by Thermospore** [comment]
Sweet! I have some feedback, now that I'm able to play with it myself:
https://youtu.be/mo0Ao9DIAqU

enumerated for easy reference:
1. remove the manual priority boxes and have that numbering on the left hand side be the priority number
2. make the priority number a rank, ie 1 = 1st = most priority (to keep the settings and search page in the same order)
3. currently when you set a new position number, the two dictionaries swap places. instead, the other dictionaries should slide out of the way (see above anki ui example)

I think this would be ideal, for the reasons explained in the vid

(I'd be willing to submit a pull request but I'm pretty sure it's beyond my ability 👀 )

**Comment by toasted-nutbread** [comment]
I'm hesitant to remove the priority number, at least right now, since there may be users that use or prefer that method, or maybe want to use both. There's really not much information available about how people use Yomichan and what features are/aren't used, which can be both good and bad.

For ordering by the index/rank, this is technically already done, but it's done at a stage that is potentially later than it should be. I could probably make the `dictionaryIndex` comparison occur just after the `dictionaryPriority` comparison and it would function as you describe, assuming all priorities are set to 0.

https://github.com/FooSoft/yomichan/blob/e7d349c3ec75f61bab09035226921968f7423741/ext/js/language/translator.js#L1186-L1228

**Comment by Thermospore** [comment]
Ohh I see so you can simply set all the priority numbers the same, making the sort rely on `dictionaryIndex`. Seems to work! That plus #1667 should cover all of the functionality I was looking for. Thanks!!
![image](https://user-images.githubusercontent.com/35350819/117896651-793aad00-b26d-11eb-8f6c-ac7f815b46bb.png)

**Comment by Thermospore** [comment]
#1667 working well so far, thanks again

**Comment by Thermospore** [comment]
> For ordering by the index/rank, this is technically already done, but it's done at a stage that is potentially later than it should be. I could probably make the dictionaryIndex comparison occur just after the dictionaryPriority comparison and it would function as you describe, assuming all priorities are set to 0.

oh whoops I didn't read that correctly the first time (just got my covid vaccine; totally out of it lol)

**Comment by Thermospore** [comment]
Yeah the grouping isn't done properly when all the priorities are set the same, with the current sort code
![image](https://user-images.githubusercontent.com/35350819/118196325-52a97d00-b3f9-11eb-8cd3-1ea3a21b4da3.png)

**Comment by toasted-nutbread** [comment]
By "grouping" do you mean "ordering"? It's expected that the ordering is not the same as it's currently designed, and I don't think that the order can be used the same as the priority, because that will cause dictionary the order of entries to be incorrect. I.e., the definitions of the first dictionary would _always_ come before any other dictionary, even if later dictionaries have definitions that match better due to a higher score. Related: #1668
